lib/decompress/zstd_decompress_block.c

```diff
@@ -189,23 +230,57 @@ size_t ZSTD_decodeLiteralsBlock(ZSTD_DCtx* dctx,
                     break;
                 }
 
+                RETURN_ERROR_IF(litSize > 0 && dst == NULL, dstSize_tooSmall, "NULL not handled");
+                RETURN_ERROR_IF(expectedWriteSize < litSize, dstSize_tooSmall, "");
+                if (streaming == not_streaming && dstCapacity > ZSTD_BLOCKSIZE_MAX + 2 * WILDCOPY_OVERLENGTH)
+                {
+                    /* room for litbuffer to fit without being stomped by output or read faulting */
+                    dctx->litBuffer = (BYTE*)dst + ZSTD_BLOCKSIZE_MAX + WILDCOPY_OVERLENGTH - litSize;
+                    dctx->litBufferEnd = dctx->litBuffer + litSize;
+                    dctx->litLocation = ZSTD_lit_in_dst;
+                }
+                else if (litSize > ZSTD_LITBUFFEREXTRASIZE)
+                {
+                    /* won't fit in litExtraBuffer, so it will be split between end of dst and extra buffer */
+                    /* unlike the huffman case, literals will be directly copied to separate buffers */
+                    dctx->litBuffer = (BYTE*)dst + expectedWriteSize - litSize + ZSTD_LITBUFFEREXTRASIZE - WILDCOPY_OVERLENGTH;
+                    dctx->litBufferEnd = dctx->litBuffer + litSize - ZSTD_LITBUFFEREXTRASIZE;
+                    dctx->litLocation = ZSTD_lit_is_split;
+                }
+                else
+                {
+                    /* fits entirely within litExtraBuffer, so no split is necessary */
+                    dctx->litBuffer = dctx->litExtraBuffer;
+                    dctx->litBufferEnd = dctx->litBuffer + litSize;
+                    dctx->litLocation = ZSTD_lit_safe;
+                }
```

Comment:
nit: This code is duplicated ~3x with only minor changes, can it be extracted into a function?